pkg/azure/provider/helpers/driver.go

```diff
@@ -285,15 +300,15 @@ func getOSDiskToUpdate(storageProfile *armcompute.StorageProfile) *armcompute.OS
 	return updatedOSDisk
 }
 
-// getDataDisksToUpdate checks if cascade delete option set for all DataDisks attached to the Virtual machine.
+// getDataDisksToUpdate checks if cascade delete option set for all DataDisks attached to the Virtual machine that were part of VM creation.
 // All data disks that do not have that set, it will set the appropriate DeleteOption and return the updated
 // DataDisks else it will return nil
-func getDataDisksToUpdate(storageProfile *armcompute.StorageProfile) []*armcompute.DataDisk {
+func getDataDisksToUpdate(storageProfile *armcompute.StorageProfile, dataDisksToUpdate []string) []*armcompute.DataDisk {
 	var updatedDataDisks []*armcompute.DataDisk
 	if storageProfile != nil && !utils.IsSliceNilOrEmpty(storageProfile.DataDisks) {
 		updatedDataDisks = make([]*armcompute.DataDisk, 0, len(storageProfile.DataDisks))
 		for _, dataDisk := range storageProfile.DataDisks {
-			if dataDisk.DeleteOption == nil || *dataDisk.DeleteOption != armcompute.DiskDeleteOptionTypesDelete {
+			if (dataDisksToUpdate != nil && slices.Contains(dataDisksToUpdate, *dataDisk.Name)) && (dataDisk.DeleteOption == nil || *dataDisk.DeleteOption != armcompute.DiskDeleteOptionTypesDelete) {
```

Comment:
Check `dataDisksToUpdate` can be done upfront and returned if its nil or empty without even entering the for loop. There is no need to repeat this check for every loop iteration.